File: Olympiad/COCI/official/2007/contest5_solutions/ivana.c

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task IVANA
  Programming language C
*/

#include <stdio.h>

#define MAXN 100

int max( int a, int b ) { return a>b ? a : b; }

int main()
{
   int n, i, L;
   static int f[2*MAXN][2*MAXN];
   int rj;

   scanf( "%d", &n );

   for ( i=0; i<n; ++i ) {
      int x;
      scanf( "%d", &x );
      x %= 2;
      f[i+n][i+n] = f[i][i] = x;
   }

   for ( L=2; L<=n; ++L )
      for ( i=0; i<2*n; ++i ) {
         int j = i+L-1;
         if ( j>=2*n ) break;

         f[i][j] = max( f[i][i] - f[i+1][j], f[j][j] - f[i][j-1] );
      }

   rj = 0;
   for ( i=0; i<n; ++i )
      if ( f[i][i] - f[i+1][i+n-1] > 0 )
         ++rj;
   printf( "%d\n", rj );

   return 0;
}

File: Olympiad/COCI/official/2007/contest5_solutions/ivana.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task IVANA
  Programming language Pascal
*)

program ivana;

const MAXN = 100;

var
   n, x, L, i, j, rj : integer;
   f                 : array[1..2*MAXN, 1..2*MAXN] of integer;

function max(a, b : integer):integer;
begin
   if a > b then
      max := a
   else
      max := b;
end;

begin
   readln(n);
   for i:=1 to n do begin
      read(x);
      f[i, i] := x mod 2;
      f[i+n, i+n] := x mod 2;
   end;

   for L:=2 to n do
      for i:=1 to 2*n do begin
         j := i+L-1;
         if j > 2*n then
            break;

         f[i, j] := max( f[i, i] - f[i+1, j], f[j, j] - f[i, j-1] );
      end;

   rj := 0;
   for i:=1 to n do
      if f[i, i] - f[i+1, i+n-1] > 0 then
         rj := rj + 1;
   writeln(rj);
end.
